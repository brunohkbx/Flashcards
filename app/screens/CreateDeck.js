import React, { Component } from 'react';
import { connect } from 'react-redux';
import { View } from 'react-native';
import DeckForm from '../components/DeckForm';
import { createDeck } from '../actions';
import { Colors, ToolbarAction } from 'react-native-paper';

export class CreateDeck extends Component {
  static navigationOptions = ({ navigation }) => {
    return {
      headerRight:
        <ToolbarAction
          icon="done"
          color={Colors.white}
          onPress={() => navigation.getParam('submitForm')()}
        />
    };
  };

  componentDidMount() {
    this.props.navigation.setParams({ submitForm: this.submitForm });
  }

  handleSubmit = (values, actions) => {
    const { createDeck } = this.props;

    createDeck(values);
    actions.setSubmitting(false);
  }

  submitForm = () => this.form.submitForm();

  render() {
    return (
      <View style={{flex: 1, padding: 8}}>
        <DeckForm
          formRef={node => this.form = node}
          handleSubmit={this.handleSubmit}
          title="Create New Deck"
          actionSubmitText="Create"
        />
      </View>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  createDeck: data => dispatch(createDeck(data))
});

export default connect(null, mapDispatchToProps)(CreateDeck);
