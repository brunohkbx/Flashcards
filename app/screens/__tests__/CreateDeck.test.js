import React from 'react';
import { shallow } from 'enzyme';
import { CreateDeck } from '../CreateDeck';

describe('CreateDeck', () => {
  const setup = propOverrides => {
    const defaultProps = Object.assign({
      navigation: {
        setParams: jest.fn()
      },
      createDeck: jest.fn()
    }, propOverrides)

    const wrapper = shallow(<CreateDeck {...defaultProps} />);
    const wrapperInstance = wrapper.instance();

    return {
      wrapper,
      wrapperInstance
    }
  }

  it('renders properly', () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it('assigns submitForm to navigations params when component did mount', () => {
    const mockNavigation = { setParams: jest.fn() }
    const { wrapperInstance } = setup({ navigation: mockNavigation });

    expect(
      mockNavigation.setParams
    ).toHaveBeenCalledWith({ submitForm: wrapperInstance.submitForm });
  });

  describe('handleSubmit', () => {
    it('calls createDeck', () => {
      const mockCreateDeck = jest.fn();
      const actions = { setSubmitting: jest.fn() };
      const { wrapperInstance } = setup({ createDeck: mockCreateDeck });

      wrapperInstance.handleSubmit({}, actions);

      expect(mockCreateDeck).toHaveBeenCalledWith({});
    });
  });

  describe('submitForm', () => {
    it('calls submitForm on formik ref', () => {
      const mockSubmitForm = jest.fn();
      const { wrapperInstance } = setup();
      wrapperInstance.form = { submitForm: mockSubmitForm };

      wrapperInstance.submitForm();

      expect(mockSubmitForm).toHaveBeenCalled();
    });
  });

  describe('setFormRef', () => {
    it('assigns a ref to this.form', () => {
      const mockForm = jest.fn();
      const { wrapperInstance } = setup();

      wrapperInstance.setFormRef(mockForm);

      expect(wrapperInstance.form).toEqual(mockForm);
    });
  });
});
