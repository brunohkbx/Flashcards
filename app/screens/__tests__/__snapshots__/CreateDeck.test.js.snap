// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateDeck renders properly 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "padding": 8,
    }
  }
>
  <FormikDeckForm
    actionSubmitText="Create"
    formRef={[Function]}
    handleSubmit={[Function]}
    initialValues={
      Object {
        "id": "606a1255-bac9-4f56-bc9e-d7b2ed696a35",
        "questions": Array [],
        "title": "",
      }
    }
    title="Create New Deck"
  />
</Component>
`;
